### ISSUE TITLE
Module panos_zone is deleting interface members

### ISSUE THREAD

**Issue by helgeolav** [PaloAltoNetworks/ansible-pan#373]
According to the examples for using this zone I can create a zone using the module without specifying interface. Then add an interface to the zone by specifying the interface. However when I run panos_zone without any interface parameter at all it deletes all members from the zone. Also if I just add one more interface to the zone, the previous are deleted. Meaning a snippet like this:

```
- name: Create zone
  panos_zone:
    name: inside
- name: Add interface to zone
  panos_zone:
   name: inside
   interface: ethernet1/1
- name: Add another interface to zone
  panos_zone:
   name: inside
   interface: ethernet1/2
# At this point the zone has only ethernet1/2 as a member
- name: Verify that the zone exist (or whatever you do later on)
  panos_zone:
    name: inside
```
This leaves me in a state where the zone has no members at all. I don't this is the expected behaviour of the module.

**Comment by shinmog** [comment]
This is exactly the expected behavior.  Ansible's `state=present` means, "I don't know what's there, but make sure that what's there looks like this."

If you are wanting to append something to pre-existing config, you need to first know what's there, then invoke the appropriate module with the new information merged in.  From what I understand, this is typically achieved with an invocation to the appropriate "facts" module first to get the current config, then you invoke a config module with the new entry appended.  There is currently no "panos_zone_facts", so we can rename this issue to a feature request for that if you want.

**Comment by ishuguru** [comment]
This is more of a general question regarding ansible and PA. How would you recommend to add config where there are no current modules? Say for example to add dhcp relay setting on a interface. Would you use the panos_op or some generic rest api module (i saw there was some new one in 2.8)?
Or somhow generate the xml file localy and then import/merge them on the device?

**Comment by helgeolav** [comment]
Thanks for the reply. At least the documentation is wrong for this module as it says that the interface parameter adds an interface but when it in reality replaces the list.

What should be done is then to move the interface management to a new module; panos_zone_interface so we can manage that separately. And also delete members. As it is now I feel it is dangerous to use as it deletes existing information. When I call panos_bgp_peer_group multiple times I don't expect all peers under that group to disappear.

@ishuguru: For now I use the URI module to do everything that is not working from ansible-pan. It would have been nice with a module like panos_op that I can use to easily do XML RPC calls.

**Comment by ishuguru** [comment]
I see
@shinmog Do you have any input on this subject?

**Comment by shinmog** [comment]
I don't think there needs to be an Ansible module for this, we just need to use the Ansible tools already there..?  Please tell me if I'm missing something, because this works just fine for me.

Here's an example updating the interfaces contained in a virtual router (in practice, this would come from a `panos_virtual_router_facts` module instead of being a static variable):

```yaml
- name: Network Playbook
  hosts: fw
  connection: local
  gather_facts: false

  vars:
    myspec:
      name: 'ansible vr'
      interface:
        - loopback.11
        - loopback.12
        - loopback.13

  roles:
    - role: PaloAltoNetworks.paloaltonetworks

  tasks:
  - name: Grab auth creds
    include_vars: 'vars.yml'
    no_log: 'yes'


  - name: Create vr
    panos_virtual_router:
      provider: '{{ aws_provider }}'
      name: '{{ myspec.name }}'
      interface: '{{ myspec.interface }}'
      commit: false

  - name: Update the vr (removing non existing interface)
    panos_virtual_router:
      provider: '{{ aws_provider }}'
      name: '{{ myspec.name }}'
      interface: '{{ myspec.interface | difference(["loopback.14"]) }}'
      commit: false

  - name: Update the vr (removing 12)
    panos_virtual_router:
      provider: '{{ aws_provider }}'
      name: '{{ myspec.name }}'
      interface: '{{ myspec.interface | difference(["loopback.12"]) }}'
      commit: false
```

Results:

```bash
(v9) $ ansible-playbook -i hosts updatevar.yml 

PLAY [Network Playbook] *****************************************************************************************************************

TASK [PaloAltoNetworks.paloaltonetworks : Install pan-python required library] **********************************************************
ok: [10.8.202.114]

TASK [PaloAltoNetworks.paloaltonetworks : Install pandevice required library] ***********************************************************
ok: [10.8.202.114]

TASK [PaloAltoNetworks.paloaltonetworks : Install xmltodict required library] ***********************************************************
ok: [10.8.202.114]

TASK [Grab auth creds] ******************************************************************************************************************
ok: [10.8.202.114]

TASK [Create vr] ************************************************************************************************************************
changed: [10.8.202.114]

TASK [Update the vr (removing non existing interface)] **********************************************************************************
ok: [10.8.202.114]

TASK [Update the vr (removing 12)] ******************************************************************************************************
changed: [10.8.202.114]

PLAY RECAP ******************************************************************************************************************************
10.8.202.114               : ok=7    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

(v9) $ 
```

As shown in my output, the removal of an interface that doesn't exist is a no-op, while removing an interface that does exist results in a change.

**Comment by ishuguru** [comment]
I dont really understand, if i just need to add a new interface to a existing zone (a task done almost daily), should i get all the current interfaces from the zone and add a new one? How is that accomplished without a panos_zone_facts module?

**Comment by shinmog** [comment]
@ishuguru 

No, you're exactly right; there would need to be a `panos_zone_facts` module.  This is what I was referring to in my first comment, and why I was asking if this defect should be renamed to a feature request for `panos_zone_facts`.

**Comment by shinmog** [comment]
Fixed by #382 - added `panod_zone_facts`
